Add zones support for vSphere cloud provider(in-tree)

jiatongw opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This PR added zones(built-in node labels) support for vSphere cloud provider(in-tree).  More details can be found in the issue as below.

**Which issue(s) this PR fixes** :
Partially fixes phase 1 of issue #64021 

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add zones support for vSphere cloud provider
```

jiatongw commented:
/assign @dougm

jiatongw commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -1306,3 +1312,106 @@ func (vs *VSphere) NodeManager() (nodeManager *NodeManager) {
+func (vs *VSphere) GetZoneByProviderID(ctx context.Context, providerID string) (cloudprovider.Zone, error) {
+
+	return cloudprovider.Zone{}, nil
+}
```
`GetZoneByNodeName()` and `GetZoneByProviderID()` are called by out-tree cloud provider, which have not been tested yet. I have completed the code for those two funcs but for now I leave it just by returning an empty zone. If this is not a good way, please comment.

jiatongw replied:
Changed to `cloudprovider.NotImplemented` here.

jiatongw commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -1325,13 +1325,13 @@ func (vs *VSphere) GetZone(ctx context.Context) (cloudprovider.Zone, error) {
 	if err != nil {
-		glog.Errorf("Cannot connent to datacenter. Get zone for node %s error", nodeName)
+		glog.Errorf("Cannot connect to datacenter. Get zone for node %s error", nodeName)
 		return cloudprovider.Zone{}, err
```
Fix some typos here

jiatongw replied:
Fix some typos here

jiatongw commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -1386,14 +1386,14 @@ func (vs *VSphere) GetZone(ctx context.Context) (cloudprovider.Zone, error) {
 			}
-			glog.Infof("No zones support for node %s error", nodeName)
+			glog.Infof("No zones support for node %s", nodeName)
 			return nil
```
Fix some typos here

jiatongw replied:
Fix some typos here

neolit123 commented:
@kubernetes/sig-cloud-provider-pr-reviews 
@kubernetes/vmware

neolit123 commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -1306,3 +1312,106 @@ func (vs *VSphere) NodeManager() (nodeManager *NodeManager) {
+	nodeName, err := vs.CurrentNodeName(ctx, vs.hostName)
+	if err != nil {
+		glog.Errorf("Cannot get node name.")
+		return cloudprovider.Zone{}, err
```
[2]

neolit123 commented on pkg/cloudprovider/providers/vsphere/vclib/datacenter.go:
```diff
@@ -82,6 +82,33 @@ func (dc *Datacenter) GetVMByUUID(ctx context.Context, vmUUID string) (*VirtualM
+	err = pc.RetrieveOne(ctx, virtualMachine.Reference(), []string{"summary.runtime.host"}, &vmMo)
+	if err != nil {
+		glog.Errorf("Failed to retrive VM runtime host, err: %v", err)
+		return nil, err
```
[2]

neolit123 commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -1306,3 +1312,106 @@ func (vs *VSphere) NodeManager() (nodeManager *NodeManager) {
+	dc, err := vclib.GetDatacenter(ctx, vsi.conn, vs.cfg.Workspace.Datacenter)
+	if err != nil {
+		glog.Errorf("Cannot connect to datacenter. Get zone for node %s error", nodeName)
+		return cloudprovider.Zone{}, err
```
[2]

neolit123 commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -1306,3 +1312,106 @@ func (vs *VSphere) NodeManager() (nodeManager *NodeManager) {
+
+	if err != nil {
+		glog.Errorf("Cannot connect to vsphere. Get zone for node %s error", nodeName)
+		return cloudprovider.Zone{}, err
```
[2]

neolit123 commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -1306,3 +1312,106 @@ func (vs *VSphere) NodeManager() (nodeManager *NodeManager) {
+	vmHost, err := dc.GetHostByVMUUID(ctx, vs.vmUUID)
+	if err != nil {
+		glog.Errorf("Cannot find VM runtime host. Get zone for node %s error", nodeName)
+		return cloudprovider.Zone{}, err
```
[2]

neolit123 commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -1306,3 +1312,106 @@ func (vs *VSphere) NodeManager() (nodeManager *NodeManager) {
+			tag, err := client.GetTag(ctx, value)
+			if err != nil {
+				glog.Errorf("Get tag %s error", value)
+				return err
```
[2]

neolit123 commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -1306,3 +1312,106 @@ func (vs *VSphere) NodeManager() (nodeManager *NodeManager) {
+			category, err := client.GetCategory(ctx, tag.CategoryID)
+			if err != nil {
+				glog.Errorf("Get category %s error", value)
+				return err
```
[2]

neolit123 reviewed (commented):
thanks for the PR @jiatongw 
i've added some comments, mostly about whitespace.

re: [2]
i see that the files use `glog.Errorf()` instead of propagating errors back to the caller which isn't exactly a good practice.
feel free to ignore this for now, but this type of glog binding is undesired and possibly needs a cleanup in a separate PR.

thanks.

dougm commented on pkg/cloudprovider/providers/vsphere/vclib/datacenter.go:
```diff
@@ -82,6 +82,33 @@ func (dc *Datacenter) GetVMByUUID(ctx context.Context, vmUUID string) (*VirtualM
+		glog.Errorf("Unable to find VM by UUID. VM UUID: %s", vmUUID)
+		return nil, ErrNoVMFound
+	}
+	virtualMachine := VirtualMachine{object.NewVirtualMachine(dc.Client(), svm.Reference()), dc}
```
Looks like you can use the existing `Datacenter.GetVMByUUID` method here instead of copy-n-paste.

dougm commented on pkg/cloudprovider/providers/vsphere/vclib/tagmanagers/tagmanegers.go:
```diff
@@ -0,0 +1,32 @@
+	return TagManagers{Client: tags.NewClient(vsURL, connection.Insecure, "")}
+}
+
+func (t TagManagers) WithClient(ctx context.Context, f func(client *tags.RestClient) error) error {
```
Typo in file name: `tagmanegers.go`

Looks like this should just be a func in vclib, rather than a new package.  And longer term will go away when vmware/govmomi#1179 is fixed.

jiatongw replied:
Ok. So it's good that I move this file into `vclib` root floder and delete `tagmanagers` folder?

dougm replied:
`func withTagsClient` in vsphere.go should be fine.  No need for the `TagManagers` type.   If you do feel there's a need for the type, let's call it `tagsClient` instead of `TagManagers`.  And no need for a separate file as this function will go away when vmware/govmomi#1179 is fixed.

dougm reviewed (commented):
@jiatongw looking good.  You'll need to restore the Godeps/LICENSES files, not sure why there are "103,568 deletions"?

jiatongw commented on pkg/cloudprovider/providers/vsphere/vclib/datacenter.go:
```diff
@@ -82,6 +82,33 @@ func (dc *Datacenter) GetVMByUUID(ctx context.Context, vmUUID string) (*VirtualM
+	svm, err := s.FindByUuid(ctx, dc.Datacenter, vmUUID, true, nil)
+	if err != nil {
+		glog.Errorf("Failed to find VM by UUID. VM UUID: %s, err: %+v", vmUUID, err)
+		return nil, err
```
thanks for catching this! But looking at other code in this file, it seems that `glog.Errorf` is using more frequently. @dougm there are 44 places that using `glog.Errof`. What do you think about this or `fmt.Errorf`?

dougm replied:
There are cases in VCP where `return fmt.Errorf` is used too.  I think its ok to leave as `glog.Errorf` as it can be useful to have the `filename:linenumber` info that it provides.  We should probably have a separate issue to clean up the the logging in VCP.  In this particular case, looks like this code should be calling `Datacenter.GetVMByUUID` instead as I commented below.

dougm commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -808,8 +816,7 @@ func (vs *VSphere) LoadBalancer() (cloudprovider.LoadBalancer, bool) {
 func (vs *VSphere) Zones() (cloudprovider.Zones, bool) {
-	glog.V(1).Info("The vSphere cloud provider does not support zones")
-	return nil, false
+	return vs, true
```
As we are discussing if vcp config should be required on the nodes, maybe we should return `false` here if there is no vcenter configured?

jiatongw replied:
you mean no `vcp.conf` file or no `[labels]` field in `vcp.conf` file? 
If it's the second case, since the kubelet code will detect whether the `zone` and `region` is `""` or not. If it's empty, the kubelet code will handle this. So we can just `return vs, true` here.

dougm replied:
I was thinking no vcenter == `if vs.cfg.Workspace.VCenterIP == "": return false`

jiatongw replied:
If there is no `vcp.conf` file, `Zones()` maybe even not called since `get node` return `No resources found`

jiatongw replied:
I see. one more concern, Should we also apply this check to other functions/interfaces in `vsphere.go` which they use `vcp.conf` file?

dougm commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -1306,3 +1313,102 @@ func (vs *VSphere) NodeManager() (nodeManager *NodeManager) {
+}
+
+func (vs *VSphere) GetZoneByNodeName(ctx context.Context, nodeName k8stypes.NodeName) (cloudprovider.Zone, error) {
+	return cloudprovider.Zone{}, nil
```
Should probably return `cloudprovider.NotImplemented` instead of `nil` here.

dougm commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -1306,3 +1313,102 @@ func (vs *VSphere) NodeManager() (nodeManager *NodeManager) {
+}
+
+func (vs *VSphere) GetZoneByProviderID(ctx context.Context, providerID string) (cloudprovider.Zone, error) {
+	return cloudprovider.Zone{}, nil
```
Should probably return `cloudprovider.NotImplemented` instead of `nil` here.

dougm commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -1306,3 +1313,102 @@ func (vs *VSphere) NodeManager() (nodeManager *NodeManager) {
+	}
+	client := vsi.conn
+	err = withTagsClient(ctx, client, func(client *tags.RestClient) error {
+		tags, err := client.ListAttachedTags(
```
This a short parameter list, no need for multiple lines: `client.ListAttachedTags(ctx, vmHost)`

Commit:
Add zones support for vSphere cloud provider (in-tree)

Commit:
Update vmware/govmomi vendor: add vapi package

Zones implementation for vSphere cloud provider needs dependencies
which are not included in current vmware/govmomi vendor. So this
update added "vapi" package to support zones.

jiatongw commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -1306,3 +1317,99 @@ func (vs *VSphere) NodeManager() (nodeManager *NodeManager) {
+	client := vsi.conn
+	err = withTagsClient(ctx, client, func(client *tags.RestClient) error {
+		tags, err := client.ListAttachedTags(ctx, vmHost)
+		if err != nil {
```
Changed to one line arguments

jiatongw replied:
And removed all not necessary blank lines, fixed problems in the above conversations.

jiatongw commented:
For those who review this PR, my changes on k8s code are the file `vsphere.go` and `vclib/datacenter.go`. The other 40 changes of file are because I updated the vmware/govmomi vendor and used [Godep](https://github.com/kubernetes/community/blob/master/contributors/devel/godep.md). So you can ignore those file changes when reviewing this PR

dougm commented on pkg/cloudprovider/providers/vsphere/vclib/datacenter.go:
```diff
@@ -82,6 +82,21 @@ func (dc *Datacenter) GetVMByUUID(ctx context.Context, vmUUID string) (*VirtualM
+		return nil, err
+	}
+	host := vmMo.Summary.Runtime.Host
+	glog.Infof("%s host is %s", virtualMachine.Reference(), vmMo.Summary.Runtime.Host)
```
You can make use of the `host` variable here too.

dougm reviewed (approved):
/ok-to-test
/approve

jiatongw commented:
/assign @cblecker

Hi @cblecker, I found that you are one of the dep-approvers. Could you fell free to review this PR? Thanks!

cblecker commented:
/ok-to-test

cblecker commented:
@jiatongw This is a legitimate test failure. Please review.

jiatongw commented:
@cblecker Yes, I have to modify the test func in `_test.go` file

Commit:
Add TestZone()

jiatongw commented:
@dougm Modified `TestZone()` in `vsphere_test.go`. We need test both `GetZone()` and `Zones()` interface here. Currently the we don't have simulator support, so the `GetZone()` will return an empty zone and nil error.  `Zones()` will return `true` because we implemented this interface.

dougm commented:
@jiatongw yes, that's good for now.  I am currently working on vcsim support for the tags API.

jiatongw commented:
@cblecker tests passed. Please have a review, thanks!

dougm commented:
/lgtm

cblecker commented:
/approve
Approving for godeps.
